# How to search
---

Before we do any work, we need to import several functions from cdapython:


- `Q` and `query` which power the search
- `columns` which lets us view entity field names
- `unique_terms` which lets view entity field contents
    
We're also asking cdapython to report it's version so we can be sure we're using the one we mean to.

In [1]:
from cdapython import Q, columns, unique_terms, query
import cdapython
print(cdapython.__version__)
Q.set_host_url("http://35.192.60.10:8080/")

2022.4.13



The CDA provides a custom python tool for searching CDA data. [`Q`](usage/#q) (short for Query) offers several ways to search and filter data, and several input modes:

---

- **[Q.counts()](../../../Documentation/usage/#qcounts)** returns summary information (counts) for the files that fit the specified search
- **[Q.files()](../../../Documentation/usage/#qfiles)** returns data for the files that fit the specified search
- **[Q.run()](../../../Documentation/usage/#qrun)** returns data for the specified search 
- **[Q.sql()](../../../Documentation/usage/#qsql)** allows you to use SQL syntax instead of Q syntax 
- **[query()](../../../Documentation/usage/#query)** allows you to use a more natural language syntax of Q

---

## Retrieving summary information

Let's run a search with a relatively simple question: We're interested how well represented hispanic and/or latino populations are in CDA resources and would like to know what data is available. Since `counts()` answers questions about data, by default it searches the "Files" table, which means we need to search with file terms. To run this simple search, we would first construct a query in `Q` and save it to a variable `myquery`:

In [21]:
myquery = Q('Subject.ethnicity = "hispanic or latino"')

---

#### Help: Where did those terms come from?
    
If you aren't sure how we knew what terms to put in our search, please refer back to the [What search terms are available?](../SearchTerms) notebook. 

---

Since we are looking for summary information, we want to use this query in `Q.counts`. We do this by running `.counts()` on the query we just saved, and saving the result to a new variable `mycounts`:

In [22]:
mycounts = myquery.counts()

(500)
Reason: 
HTTP response headers: HTTPHeaderDict({'Content-Type': 'application/json', 'Transfer-Encoding': 'chunked', 'Date': 'Thu, 21 Apr 2022 19:05:53 GMT', 'Connection': 'close'})
HTTP response body: {"message":null,"statusCode":500,"causes":[null]}



We're saving information in variables, so we don't get any visible output. To see what our results are, we need to look into the variable. The simplest way is to call `mycounts` directly:

In [4]:
mycounts


            QueryID: 178f586e-4a4a-4795-b4b5-aec35082fc79
            
            Offset: 0
            Count: 3
            Total Row Count: 3
            More pages: False
            

This output tells us our QueryID, which we don't really need, but the computer does to track our questions. Then it tells us four parameters that describe our results:

---

- **Offset:** This is how many rows of information we've told the query to skip in the data, here we didn't tell it to skip anything, so the offset is zero
- **Count:** This is how many rows the current page of our results table has. To keep searches fast, we default to pages with 100 rows.
- **Total Row Count:** This is how many rows are in the full results table
- **More pages:** This is alwasys a True or False. False means that our current page has all the availble results. True means that we will see only the first 100 results in this table, and will need to page through for more.

---
    
Now that we've seen the metadata about our results, let's look at the actual table. The easiest way to do this is by using the python function `.to_dataframe()` on our `mycounts` variable:

In [5]:
mycounts.to_dataframe()

system subject_count subject_files_count researchsubject_count  \
0    IDC           371               86776                     0   
1    PDC            57                9401                   444   
2    GDC          3118               32866                 19158   

  researchsubject_files_count specimen_count specimen_files_count  
0                           0              0                    0  
1                        9401           1033                 9232  
2                       32866          20598                30995

### What do these numbers mean?

---
    
- **system:** Which data source contributed this data? The CDA currently has data from IDC, PDC and GDC
- **subject_count:** How many unique individuals meet our query. Note that *within* a data source the number is of *unique* individuals, but the same individuals can have data at multiple centers. Here, there are 371 unique people in the IDC data, however up to 57 of those may be exactly the same people as are in the PDC data.
- **subject_files_count:** This tells you roughly how much data is available. It is the total count of files for all the subjects in `subject_count`, which is also the total number of files that match your search.
- **researchsubject_count:** Some data sources have individual subjects that are in multiple studies, when this happens the individual will have both a "subject" identifier and a "researchsubject" identifier. This column counts the latter. Zero in this column can mean either "there are no research_subjects that meet your search criteria" or "the data source for this row does not create special identifiers for subjects in multiple studies"
- **researchsubject_files_count:** This is the total count of files specific to researchsubjects in `researchsubject_count`. It is a subset of `subject_count`
- **specimen_count:** Some data sources track whether files come from specific specimens from a given individual. This column counts the number of specimens that meet your search criteria. Zero in this column can mean either "there are no specimens that meet your search criteria" or "the data source for this row does not track specimens seperately from subjects"
- **specimen_files_count:** This is the total count of files specific to specimens in `specimen_count`. It is a subset of both `subject_count` and `researchsubject_files_count`

---


If you are comfortable with python and/or don't need the intermediate variables for other queries, you can string all of those same commands together to get the output as a table directly from `Q`:

In [6]:
Q('Subject.ethnicity = "hispanic or latino"').counts().to_dataframe()



system subject_count subject_files_count researchsubject_count  \
0    IDC           371               86776                     0   
1    PDC            57                9401                   444   
2    GDC          3118               32866                 19158   

  researchsubject_files_count specimen_count specimen_files_count  
0                           0              0                    0  
1                        9401           1033                 9232  
2                       32866          20598                30995

## Retrieving data


Let's run the same query, but instead of asking for summary information, lets get the actual data. We start this process the same way, by making a `Q` statement. We can reuse `myquery` from earlier, and just run the `.files()` function on it: 

In [7]:
mydata = myquery.files()

Getting results from database



In [23]:
mydata


            QueryID: 2acc5aa0-386b-40d7-9d2b-a41f5b980e9e
            Query:SELECT all_v3_0_Files.* FROM gdc-bq-sample.dev.all_v3_0_Files AS all_v3_0_Files, UNNEST(all_v3_0_Files.Subject) AS _Subject WHERE (UPPER(_Subject.ethnicity) = UPPER('hispanic or latino'))
            Offset: 0
            Count: 100
            Total Row Count: 139124
            More pages: True
            

To see the data, we again put the result in a dataframe. This would normally show 100 rows of data, so we use `.head()` to only see the first 10:

In [9]:
mydata.to_dataframe().head()

id  \
0  780c44bb-3783-460f-862d-074a70a89941   
1  4454d6eb-8638-46fd-bbd5-0af0457177a7   
2  7d1ce6ef-c0f2-4707-a95a-8eebbe50963a   
3  d0859f49-775b-4fc9-8281-310dbfef7d93   
4  8aa97acb-d8ab-48c1-b9e2-d90c2506ad31   

                                          identifier  \
0  [{'system': 'GDC', 'value': '780c44bb-3783-460...   
1  [{'system': 'PDC', 'value': '4454d6eb-8638-46f...   
2  [{'system': 'GDC', 'value': '7d1ce6ef-c0f2-470...   
3  [{'system': 'GDC', 'value': 'd0859f49-775b-4fc...   
4  [{'system': 'GDC', 'value': '8aa97acb-d8ab-48c...   

                                               label  \
0  CAUSE_p_TCGAb73_SNP_N_GenomeWideSNP_6_D07_6982...   
1  TCGA_114C_29-1696-01A-01_29-1771-01A-01_13-206...   
2  TARGET_AML.18578b8f-7557-4876-8c68-a864611bcaf...   
3                   D0CM7ACXX_3_ATTGGC_gdc_realn.bam   
4  fe619ca3-42d5-4281-9b29-0911b982b877_mirnaseq_...   

                 data_category                   data_type file_format  \
0        Copy Number Variation  Masked Copy Number Segment         TXT   
1       Processed Mass Spectra               Open Standard        mzML   
2  Simple Nucleotide Variation  Annotated Somatic Mutation         VCF   
3             Sequencing Reads               Aligned Reads         BAM   
4             Sequencing Reads               Aligned Reads         BAM   

  associated_project                                            drs_uri  \
0          TCGA-UCEC  drs://dg.4DFC:780c44bb-3783-460f-862d-074a70a8...   
1         CPTAC-TCGA  drs://dg.4DFC:4454d6eb-8638-46fd-bbd5-0af04571...   
2         TARGET-AML  drs://dg.4DFC:7d1ce6ef-c0f2-4707-a95a-8eebbe50...   
3         TARGET-AML  drs://dg.4DFC:d0859f49-775b-4fc9-8281-310dbfef...   
4         TARGET-AML  drs://dg.4DFC:8aa97acb-d8ab-48c1-b9e2-d90c2506...   

   byte_size                          checksum data_modality imaging_modality  \
0       5586  c19b70ae36770c902672a308609b084e       Genomic             None   
1   24717292  4391ffc0326a5d1ad70f4c552e5f4242     Proteomic             None   
2    1620816  947b90503740257bc6f29fcd364d1b3f       Genomic             None   
3  258859204  a90f970ba1540b2a033135ecef260d5f       Genomic             None   
4  146160867  53ba7995f7fa227275333210ab0316fc       Genomic             None   

  dbgap_accession_number                                            Subject  \
0                   None  [{'id': 'TCGA-BS-A0TA', 'identifier': [{'syste...   
1                   None  [{'id': 'TCGA-29-1696', 'identifier': [{'syste...   
2              phs000465  [{'id': 'TARGET-20-PARGVC', 'identifier': [{'s...   
3              phs000465  [{'id': 'TARGET-20-PARTXH', 'identifier': [{'s...   
4              phs000465  [{'id': 'TARGET-20-PAWNGC', 'identifier': [{'s...   

                                     ResearchSubject  \
0  [{'id': 'ad3d8420-f562-4cbe-ac82-c968bd4f782d'...   
1  [{'id': 'b8ebbb4c-63d7-11e8-bcf1-0a2705229b82'...   
2  [{'id': '0bb51ed6-41bb-56a3-bd80-a297992b61c9'...   
3  [{'id': '51af485c-dfcb-5e00-a22a-0d7d539892ad'...   
4  [{'id': '98e681e3-b756-4535-8b95-09f16ff66e8f'...   

                                            Specimen  
0  [{'id': 'b6a042b6-f94a-4acc-98cb-a1a2fe72b9ad'...  
1  [{'id': 'c049099a-641e-11e8-bcf1-0a2705229b82'...  
2  [{'id': '1de5bf90-45c7-5962-ab0e-29e4bd64a7df'...  
3  [{'id': 'aacb5151-785c-57ec-9129-419091dc2820'...  
4  [{'id': '255615a2-1715-4df5-9115-51b2448e8aa0'...

Just as before, we can also skip the multiple calls and just string our command together for the same answer:

In [10]:
Q('Subject.ethnicity = "hispanic or latino"').files().to_dataframe().head()



Getting results from database



id  \
0  780c44bb-3783-460f-862d-074a70a89941   
1  4454d6eb-8638-46fd-bbd5-0af0457177a7   
2  7d1ce6ef-c0f2-4707-a95a-8eebbe50963a   
3  d0859f49-775b-4fc9-8281-310dbfef7d93   
4  8aa97acb-d8ab-48c1-b9e2-d90c2506ad31   

                                          identifier  \
0  [{'system': 'GDC', 'value': '780c44bb-3783-460...   
1  [{'system': 'PDC', 'value': '4454d6eb-8638-46f...   
2  [{'system': 'GDC', 'value': '7d1ce6ef-c0f2-470...   
3  [{'system': 'GDC', 'value': 'd0859f49-775b-4fc...   
4  [{'system': 'GDC', 'value': '8aa97acb-d8ab-48c...   

                                               label  \
0  CAUSE_p_TCGAb73_SNP_N_GenomeWideSNP_6_D07_6982...   
1  TCGA_114C_29-1696-01A-01_29-1771-01A-01_13-206...   
2  TARGET_AML.18578b8f-7557-4876-8c68-a864611bcaf...   
3                   D0CM7ACXX_3_ATTGGC_gdc_realn.bam   
4  fe619ca3-42d5-4281-9b29-0911b982b877_mirnaseq_...   

                 data_category                   data_type file_format  \
0        Copy Number Variation  Masked Copy Number Segment         TXT   
1       Processed Mass Spectra               Open Standard        mzML   
2  Simple Nucleotide Variation  Annotated Somatic Mutation         VCF   
3             Sequencing Reads               Aligned Reads         BAM   
4             Sequencing Reads               Aligned Reads         BAM   

  associated_project                                            drs_uri  \
0          TCGA-UCEC  drs://dg.4DFC:780c44bb-3783-460f-862d-074a70a8...   
1         CPTAC-TCGA  drs://dg.4DFC:4454d6eb-8638-46fd-bbd5-0af04571...   
2         TARGET-AML  drs://dg.4DFC:7d1ce6ef-c0f2-4707-a95a-8eebbe50...   
3         TARGET-AML  drs://dg.4DFC:d0859f49-775b-4fc9-8281-310dbfef...   
4         TARGET-AML  drs://dg.4DFC:8aa97acb-d8ab-48c1-b9e2-d90c2506...   

   byte_size                          checksum data_modality imaging_modality  \
0       5586  c19b70ae36770c902672a308609b084e       Genomic             None   
1   24717292  4391ffc0326a5d1ad70f4c552e5f4242     Proteomic             None   
2    1620816  947b90503740257bc6f29fcd364d1b3f       Genomic             None   
3  258859204  a90f970ba1540b2a033135ecef260d5f       Genomic             None   
4  146160867  53ba7995f7fa227275333210ab0316fc       Genomic             None   

  dbgap_accession_number                                            Subject  \
0                   None  [{'id': 'TCGA-BS-A0TA', 'identifier': [{'syste...   
1                   None  [{'id': 'TCGA-29-1696', 'identifier': [{'syste...   
2              phs000465  [{'id': 'TARGET-20-PARGVC', 'identifier': [{'s...   
3              phs000465  [{'id': 'TARGET-20-PARTXH', 'identifier': [{'s...   
4              phs000465  [{'id': 'TARGET-20-PAWNGC', 'identifier': [{'s...   

                                     ResearchSubject  \
0  [{'id': 'ad3d8420-f562-4cbe-ac82-c968bd4f782d'...   
1  [{'id': 'b8ebbb4c-63d7-11e8-bcf1-0a2705229b82'...   
2  [{'id': '0bb51ed6-41bb-56a3-bd80-a297992b61c9'...   
3  [{'id': '51af485c-dfcb-5e00-a22a-0d7d539892ad'...   
4  [{'id': '98e681e3-b756-4535-8b95-09f16ff66e8f'...   

                                            Specimen  
0  [{'id': 'b6a042b6-f94a-4acc-98cb-a1a2fe72b9ad'...  
1  [{'id': 'c049099a-641e-11e8-bcf1-0a2705229b82'...  
2  [{'id': '1de5bf90-45c7-5962-ab0e-29e4bd64a7df'...  
3  [{'id': 'aacb5151-785c-57ec-9129-419091dc2820'...  
4  [{'id': '255615a2-1715-4df5-9115-51b2448e8aa0'...

What are all these fields?

---

- **id:** The unique identifier for this file
- **identifier:** An embedded array of information that includes the originating data center and the ID the file had there
- **label:** The full name of the file
- **data_catagory:** A desecription of the kind of general kind data the file holds
- **data_type:** A more specific descripton of the data type
- **file_format:** The extension of the file
- **associated_project:** The name the data center uses for the study this file was generated for
- **drs_uri:** A unique identifier that can be used to retreive this specific file from a server
- **byte_size:** Size of the file in bytes
- **checksum:** The md5 value for the file
- **data_modality:** A high level descriptor of file data, always one of "Genomic", "Proteomic", or "Imaging"
- **imaging_modality** For files with the `data_modality` of "Imaging", a descriptor for the image type
- **dbgap_accession_number:**
    
---


One thing you may have noticed is that This output shows us that there are 139124 rows of data, but this is more than the 129043 we expected. This is because `.files` has to combine Subject and File data to answer our query, and in the join, some of the files are found twice. We can  

In [11]:
unique_terms("ResearchSubject.Specimen.source_material_type", system="PDC")

['Blood Derived Normal',
 'Cell Lines',
 'Metastatic',
 'Next Generation Cancer Model',
 'Normal',
 'Normal Adjacent Tissue',
 'Not Reported',
 'Primary Tumor',
 'Recurrent Tumor',
 'Solid Tissue Normal',
 'Tumor',
 'Xenograft Tissue']

## Basic querying

We can start by getting the record for ```id = TCGA-E2-A10A``` that we mentioned earlier:

We see that we've got a single subject record as a result, which is what we expect.

Let's see how the result looks like:

In [12]:
q = Q('id = "TCGA-E2-A10A"') # note the double quotes for the string value

r = q.run()

r

Getting results from database

Total execution time: 3299 ms



            QueryID: db5dcb17-a70b-414d-a15e-b98c27adfb46
            Query:SELECT all_v3_0_subjects_meta.* FROM gdc-bq-sample.dev.all_v3_0_subjects_meta AS all_v3_0_subjects_meta WHERE (UPPER(all_v3_0_subjects_meta.id) = UPPER('TCGA-E2-A10A'))
            Offset: 0
            Count: 1
            Total Row Count: 1
            More pages: False
            

In [13]:
stage = Q('ResearchSubject.Diagnosis.stage = "Stage IV"')
r = stage.run()
print(r)

Getting results from database

Total execution time: 4207 ms

            QueryID: c08d8dad-ba2e-4d72-a78e-8261937b6b36
            Query:SELECT all_v3_0_subjects_meta.* FROM gdc-bq-sample.dev.all_v3_0_subjects_meta AS all_v3_0_subjects_meta, UNNEST(all_v3_0_subjects_meta.ResearchSubject) AS _ResearchSubject, UNNEST(_ResearchSubject.Diagnosis) AS _ResearchSubject_Diagnosis WHERE (UPPER(_ResearchSubject_Diagnosis.stage) = UPPER('Stage IV'))
            Offset: 0
            Count: 100
            Total Row Count: 100
            More pages: True
            


In [14]:
r.to_dataframe()


id                                         identifier  \
0   C3N-02333  [{'system': 'GDC', 'value': 'C3N-02333'}, {'sy...   
1     02OV041  [{'system': 'GDC', 'value': '02OV041'}, {'syst...   
2   C3N-03226  [{'system': 'GDC', 'value': 'C3N-03226'}, {'sy...   
3   C3N-01220  [{'system': 'GDC', 'value': 'C3N-01220'}, {'sy...   
4   C3N-04273  [{'system': 'GDC', 'value': 'C3N-04273'}, {'sy...   
..        ...                                                ...   
95    04OV037  [{'system': 'GDC', 'value': '04OV037'}, {'syst...   
96  C3N-01758  [{'system': 'GDC', 'value': 'C3N-01758'}, {'sy...   
97  C3N-01948  [{'system': 'GDC', 'value': 'C3N-01948'}, {'sy...   
98  C3N-04152  [{'system': 'GDC', 'value': 'C3N-04152'}, {'sy...   
99  C3N-00302  [{'system': 'GDC', 'value': 'C3N-00302'}, {'sy...   

         species     sex          race               ethnicity days_to_birth  \
0   Homo sapiens    male  not reported            not reported        -22285   
1   Homo sapiens  female         white  not hispanic or latino          None   
2   Homo sapiens    male  not reported            not reported        -19938   
3   Homo sapiens    male  not reported            not reported        -18748   
4   Homo sapiens    male  not reported            not reported        -26874   
..           ...     ...           ...                     ...           ...   
95  Homo sapiens  female         white  not hispanic or latino          None   
96  Homo sapiens    male  not reported            not reported        -21708   
97  Homo sapiens  female  not reported            not reported        -20359   
98  Homo sapiens    male  not reported            not reported        -19705   
99  Homo sapiens    male  not reported            not reported        -21532   

                  subject_associated_project  vital_status age_at_death  \
0                [CPTAC3-Discovery, CPTAC-3]         Alive         None   
1                                  [CPTAC-2]  Not Reported         None   
2   [CPTAC3-Discovery, CPTAC-3, cptac_hnscc]         Alive         None   
3   [CPTAC3-Discovery, cptac_ccrcc, CPTAC-3]          Dead          633   
4                [CPTAC3-Discovery, CPTAC-3]          Dead          425   
..                                       ...           ...          ...   
95                                 [CPTAC-2]  Not Reported         None   
96               [CPTAC3-Discovery, CPTAC-3]          Dead          712   
97  [CPTAC3-Discovery, CPTAC-3, cptac_hnscc]         Alive         None   
98  [CPTAC3-Discovery, CPTAC-3, cptac_hnscc]         Alive         None   
99    [CPTAC3-Discovery, CPTAC-3, cptac_pda]          Dead          323   

    cause_of_death                                    ResearchSubject  
0             None  [{'id': '30d5de9e-822e-4794-9a3e-cf1bccf1dab7'...  
1     Not Reported  [{'id': 'b9337dee-3bdb-49b2-90e4-97249e30418d'...  
2             None  [{'id': '42f5bee3-8217-4d9e-ad2b-42ba0ffeceae'...  
3   Cancer Related  [{'id': 'c6c75c5d-b3c9-4918-8dc2-02849933cd3c'...  
4   Cancer Related  [{'id': '701129b7-d95a-4fac-a9be-0d0da0642a81'...  
..             ...                                                ...  
95    Not Reported  [{'id': '1c545cda-2a3a-41bd-9f99-e3712e31ecdd'...  
96  Cancer Related  [{'id': 'b7eed034-4280-40fc-ae0a-d5349ae260bb'...  
97            None  [{'id': 'fa75026c-662b-422d-bde0-de3155671dc0'...  
98            None  [{'id': '91b0cc01-305b-415c-ae56-d3262780d99e'...  
99  Cancer Related  [{'id': 'ad265919-d589-4b8a-a9fb-77cca95c59f7'...  

[100 rows x 12 columns]

In [15]:
q = Q('Subject.id = "TCGA-E2-A10A"') # note the double quotes for the string value

r = q.counts(table="gdc-bq-sample.dev", version="all_v3_0_Files")

r


            QueryID: 26cf288d-401c-4bad-a48c-9691c7c4c63c
            
            Offset: 0
            Count: 3
            Total Row Count: 3
            More pages: False
            

In [16]:
r.to_dataframe()


system subject_count subject_files_count researchsubject_count  \
0    IDC             1                  15                     0   
1    GDC             1                  66                  1098   
2    PDC             1                 152                     4   

  researchsubject_files_count specimen_count specimen_files_count  
0                           0              0                    0  
1                          66             20                   44  
2                         152              8                  152

The record is pretty large, so we'll print out ```identifier``` values for each ```ResearchSubject``` to confirm that we have one ```ResearchSubject``` that comes from GDC, and one that comes from PDC:

In [17]:
for research_subject in r['system']:
    print(research_subject['system'])

AttributeError: 'str' object has no attribute 'indices'

The values represent ```ResearchSubject``` IDs and are equivalent to ```case_id``` values in data nodes.

## Example queries

Now that we can create a query with ```Q()``` function, let's see how we can combine multiple conditions.

There are three operators available:
* ```And()```
* ```Or()```
* ```From()```

The following examples show how those operators work in practice.

### Query 1

**Find data for subjects who were diagnosed after the age of 50 and who were investigated as part of the TCGA-OV project.**

In [ ]:
q1 = Q('sex = "male"')
q2 = Q('identifier.system = "GDC"')
q3 = Q('identifier.system = "PDC"')

r2 = q1.And(q2)
r3 = q1.And(q3)

#taco = r2.run()
#taco

burrito = r3.run()


In [ ]:
burrito

In [ ]:
q1 = Q('sex = "male"')

taco = q1.run()
taco

In [ ]:
from pandas import DataFrame,concat
q = query('ResearchSubject.primary_diagnosis_condition LIKE "Lung%" AND sex = "male"').run()
print(q)
df = DataFrame()
for i in q.paginator(to_df=True):
    print(len(i))
    df = concat([df,i])

In [ ]:
df

In [ ]:
r.to_dataframe()

### Query 2

**Find data for donors with melanoma (Nevi and Melanomas) diagnosis and who were diagnosed before the age of 30.**

In [ ]:
q1 = Q('ResearchSubject.Specimen.primary_disease_type = "Nevi and Melanomas"')
q2 = Q('ResearchSubject.Diagnosis.age_at_diagnosis < 30*365')

q = q1.And(q2)
r = q.run()

print(r)

In addition, we can check how many records come from particular systems by adding one more condition to the query:

In [ ]:
q1 = Q('ResearchSubject.Specimen.primary_disease_type = "Nevi and Melanomas"')
q2 = Q('ResearchSubject.Diagnosis.age_at_diagnosis < 30*365')
q3 = Q('ResearchSubject.Specimen.identifier.system = "GDC"')

q = q1.And(q2.And(q3))
r = q.run()

print(r)

By comparing the ```Count``` value of the two results we can see that all the patients returned in the initial query are coming from the GDC.

To explore the results further, we can fetch the patient JSON objects by iterating through the results:

In [ ]:
projects = set()

for patient in r:
    research_subjects = patient['ResearchSubject']
    for rs in research_subjects:
        projects.add(rs['member_of_research_project'])

print(projects)

The output shows the projects where _Nevi and Melanomas_ cases appear.

### Query 3

**Identify all samples that meet the following conditions:**

* **Sample is from primary tumor**
* **Disease is ovarian or breast cancer**
* **Subjects are females under the age of 60 years**

In [ ]:
tumor_type = Q('ResearchSubject.Specimen.source_material_type = "Primary Tumor"')
disease1 = Q('ResearchSubject.primary_diagnosis_site = "Ovary"')
disease2 = Q('ResearchSubject.primary_diagnosis_site = "Breast"')
demographics1 = Q('sex = "female"')
demographics2 = Q('days_to_birth > -60*365') # note that days_to_birth is a negative value

q1 = tumor_type.And(demographics1.And(demographics2))
q2 = disease1.Or(disease2)
q = q1.And(q2)

r = q.run()
print(r)

In this case, we have a result that contains more than 1000 records which is the default page size. To load the next 1000 records, we can use the ```next_page()``` method:

In [ ]:
r2 = r.next_page()

In [ ]:
print(r2)

Alternatively, we can use the ```offset``` argument to specify the record to start from:

```
...
r = q.run(offset=1000)
print(r)
```

### Query 4

**Find data for donors with "Ovarian Serous Cystadenocarcinoma" with proteomic and genomic data.**

**Note that disease type value denoting the same disease groups can be completely different within different systems. This is where CDA features come into play.** We first start by exploring the values available for this particular field in both systems.

In [ ]:
unique_terms('ResearchSubject.primary_diagnosis_condition', system="GDC")

Since “Ovarian Serous Cystadenocarcinoma” doesn’t appear in GDC values we decide to look into the PDC:

In [ ]:
unique_terms('ResearchSubject.primary_diagnosis_condition', system="PDC")

After examining the output, we see that it does come from the PDC. Hence, if we could first identify the data that has research subjects found within the PDC that have this particular disease type, and then further narrow down the results to include only the portion of the data that is present in GDC, we could get the records that we are looking for.

In [ ]:
q1 = Q('ResearchSubject.primary_diagnosis_condition = "Ovarian Serous Cystadenocarcinoma"')
q2 = Q('ResearchSubject.identifier.system = "PDC"')
q3 = Q('ResearchSubject.identifier.system = "GDC"')

q = q3.From(q1.And(q2))
r = q.run()

print(r)

As you can see, this is achieved by utilizing ```From``` operator. The ```From``` operator allows us to create queries from results of other queries. This is particularly useful when working with conditions that involve a single field which can take multiple different values for different items in a list that is being part of, e.g. we need ```ResearchSubject.identifier.system``` to be both “PDC” and “GDC” for a single patient. In such cases, ```And``` operator can’t help because it will return those entries where the field takes both values, which is zero entries.

In [ ]:
for i in Q.sql("SELECT * FROM `gdc-bq-sample.cda_mvp.INFORMATION_SCHEMA.COLUMN_FIELD_PATHS` WHERE table_name = 'v3' Limit 5"):
    print(i)

In [ ]:
q1 = query('ResearchSubject.identifier.system = "GDC" FROM ResearchSubject.primary_diagnosis_condition = "Ovarian Serous Cystadenocarcinoma" AND ResearchSubject.identifier.system = "PDC"')
result = q1.run(async_call=True)
print(result)


## Data extraction and release information

In [ ]:
# If you are interested in the extraction dates or data release versions of GDC, PDC, or IDC that is in a table or view, execute this code

for i in Q.sql("SELECT option_value FROM `gdc-bq-sample.integration.INFORMATION_SCHEMA.TABLE_OPTIONS` WHERE table_name = 'all_v1'"):
    print(i)